app/routes/funding-options/land-primary-actions.js

```diff
@@ -1,32 +1,90 @@
 const joi = require('joi')
-const ViewModel = require('./models/land-primary-actions')
+const { sendAgreementCalculateMessage } = require('../../messaging')
 const cache = require('../../cache')
+const { getParcels } = require('../../api/map')
 
-module.exports = [{
-  method: 'GET',
-  path: '/funding-options/land-primary-actions',
-  options: {
-    handler: async (request, h) => {
-      const agreement = await cache.get('agreement', request.yar.id)
-      return h.view('funding-options/land-primary-actions', new ViewModel(agreement.landInHectares))
+function getLandInHectares (payload, parcels) {
+  return Object.entries(payload).map(entry => {
+    const [name, value] = entry
+    const [sheetId, parcelId] = name.split('_')
+    const feature = parcels.features.find(feature => {
+      const props = feature.properties
+      return props.sheet_id === sheetId && props.parcel_id === parcelId
+    })
+
+    return {
+      name,
+      value,
+      valid: value <= feature.properties.area_ha
     }
-  }
-},
-{
-  method: 'POST',
-  path: '/funding-options/land-primary-actions',
-  options: {
-    validate: {
-      payload: joi.object({
-        landInHectares: joi.number().greater(0).required()
-      }),
-      failAction: async (request, h, error) => {
-        return h.view('funding-options/land-primary-actions', new ViewModel(request.payload.landInHectares, error)).code(400).takeover()
+  })
+}
+
+module.exports = [
+  {
+    method: 'GET',
+    path: '/funding-options/land-primary-actions',
+    options: {
+      handler: async (request, h) => {
+        const agreement = await cache.get('agreement', request.yar.id)
+        const { parcels } = await getParcels(agreement.sbi)
+
+        return h.view('funding-options/land-primary-actions', {
+          landInHectares: agreement.landInHectares,
+          landParcels: parcels
+        })
+      }
+    }
+  },
+  {
+    method: 'POST',
+    path: '/funding-options/land-primary-actions',
+    options: {
+      handler: async (request, h) => {
+        const { payload } = request
+        const agreement = await cache.get('agreement', request.yar.id)
+        const { parcels } = await getParcels(agreement.sbi)
+        const landInHectares = getLandInHectares(payload, parcels)
+        const hasInvalidValues = landInHectares.some(element => !element.valid)
+
+        if (hasInvalidValues) {
+          return h.view('funding-options/land-primary-actions', {
+            landInHectares: landInHectares,
+            landParcels: parcels,
+            errors: true
+          }).code(400).takeover()
```

Comment:
Fareed, I'll add a separate ticket to JIRA to discuss returning 400 status codes for HTML forms with errors. Personally, I don't think we should be doing this - it should be 200. I can see it is in other places though so we'll address them all together.

Reply:
Cool, no worries